### DISCUSSION TITLE
I can’t play it after recording, how can I solve it?

### DISCUSSION THREAD

**Discussion by jkloip** [flet-dev/flet#2889] [Q&A] [answered]
### Question

I use **AudioRecorder** to record and **Audio** to play the recording file. I package the test program into an apk and install it on my phone, but **_I can't hear my recording file_**. I don't know how I can solve this problem? My packaging instructions The packaging result for "**flet build apk --include-packages flet_audio_recorder flet_audio**" is normal.

### Code sample

```python
import flet as ft
path = "assets/talk/test-audio-file.wav"
url = path
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder test"), center_title=True)
    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)
    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=url,
        autoplay=False,
        volume=10,
        balance=0,
    )
    page.overlay.append(audio1)
    await page.update_async()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play", on_click=lambda _: audio1.play()),
    )

ft.app(target=main)
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by jkloip** [comment]
I am still trying to solve this problem. In the process, I tried to package AudioRecorder and Audio into a windows app, but the windows app also failed to record and had an error message:

env_openai3) PS D:\note\env_openai3\turnover2> flet build windows --include-packages flet_audio_recorder flet_audio
Creating Flutter bootstrap project...OK
Customizing app icons and splash images...OK
Generating app icons...OK
Packaging Python app...OK
Building Windows app...OK
Copying build to build\windows directory...Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "D:\note\env_openai3\Scripts\flet.exe\__main__.py", line 7, in <module>
  File "D:\note\env_openai3\Lib\site-packages\flet\cli\cli.py", line 83, in main
    args.handler(args)
  File "D:\note\env_openai3\Lib\site-packages\flet\cli\commands\build.py", line 763, in handle
    copy_tree(build_output_dir, str(out_dir), ignore=ignore_build_output)
  File "D:\note\env_openai3\Lib\site-packages\flet_runtime\utils\__init__.py", line 188, in copy_tree
    return shutil.copytree(src, dst, ignore=ignore, symlinks=True, dirs_exist_ok=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Py\Python3114\Lib\shutil.py", line 561, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Py\Python3114\Lib\shutil.py", line 515, in _copytree
    raise Error(errors)
shutil.Error: [('C:\\Users\\jkloi\\AppData\\Local\\Temp\\flet_flutter_build_zqQKxMHyLk\\build\\windows\\x64\\runner\\Release\\python3.dll', 'D:\\Note\\env_openai3\\turnover2\\build\\windows\\python3.dll', '[WinError 87] 參數錯誤。'), ('C:\\Users\\jkloi\\AppData\\Local\\Temp\\flet_flutter_build_zqQKxMHyLk\\build\\windows\\x64\\runner\\Release\\python311.dll', 'D:\\Note\\env_openai3\\turnover2\\build\\windows\\python311.dll', '[WinError 87] 參數錯誤。')]

My python version is 3.11.4 and flet version is 0.21.1
What parts should I modify so that the program can record and play recording files smoothly on Android phone? It makes me think hard.

**Comment by ndonkoHenri** [comment]
Can you remove "assets" from your path as shown [here](https://flet.dev/docs/controls/image#src) and retry?

Btw, does the app work locally (without packaging)?

**Reply by jkloip** [reply]
Thanks for your reply. I removed "assets" from the data box and changed the path to "test-audio-file.wav". After testing the recording function, the recording file was generated (the recording file could be generated before), but there was no recording. to any sound. The following message appears in command mode:
(env_openai3) PS D:\note\env_openai3\turnover2> flet .\test031702.py
D:\Note\env_openai3\turnover2\test031702.py:18: DeprecationWarning: Call to update_async() is deprecated since version 0.21.0 and will be removed in version 1.0. Use update() method instead.
  await page.update_async()
D:\Note\env_openai3\turnover2\test031702.py:27: DeprecationWarning: Call to update_async() is deprecated since version 0.21.0 and will be removed in version 1.0. Use update() method instead.
  await page.update_async()
D:\Note\env_openai3\turnover2\test031702.py:29: DeprecationWarning: Call to add_async() is deprecated since version 0.21.0 and will be removed in version 1.0. Use add() method instead.
  await page.add_async(
D:\Note\env_openai3\turnover2\test031702.py:9: DeprecationWarning: Call to start_recording_async() is deprecated since version 0.21.0 and will be removed in version 1.0. Use start_recording() method instead.
  await audio_rec.start_recording_async(path)

My code is as follows:
import flet as ft
path = "test-audio-file.wav"
url = path
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder test"), center_title=True)
    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)
    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=url,
        autoplay=False,
        volume=10,
        balance=0,
    )
    page.overlay.append(audio1)
    await page.update_async()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play", on_click=lambda _: audio1.play()),
    )

ft.app(target=main)

**Reply by jkloip** [reply]
I tried to modify the program according to the message and execute it, but the revised program was even worse because the recording function could not be used after execution. After pressing the recording button, it would jump directly from the window screen.

The revised code is as follows：
import flet as ft
path = "test-audio-file.wav"
url = path
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder test"), center_title=True)
    async def handle_start_recording(e):
        audio_rec.start_recording(path)   # before await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)
    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    page.update()    # before await page.update_async()

    audio1 = ft.Audio(
        src=url,
        autoplay=False,
        volume=10,
        balance=0,
    )
    page.overlay.append(audio1)
    page.update()    # before await page.update_async()

    page.add(
    # before await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play", on_click=lambda _: audio1.play()),
    )

ft.app(target=main)

**Reply by jkloip** [reply]
I tested it with a website example a few days ago. At that time, it was packaged into an apk file and installed on the mobile phone. The sound could be recorded and played smoothly, but this time it was indeed impossible to record and play the sound. I am trying to remember. What code adjustments were made at that time? All I remember is that I changed the recording file storage path from "test-audio-file.wav" to "assets/talk/test-audio-file.wav". But now this method is invalid!

**Reply by jkloip** [reply]
I updated the flet version to **0.21.2**, but after packaging the code into an **apk** and installing it on my phone, **AudioRecorder** still cannot record_(Test-audio-file.wav is generated on the Windows development platform, but there is no sound in it.)_. I don't know how I should solve this problem? What did I do wrong?

My code is as follows:
import flet as ft
path = "assets/talk/test-audio-file.wav"
url = path
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder test"), center_title=True)
    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)
    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=url,
        autoplay=False,
        volume=10,
        balance=0,
    )
    page.overlay.append(audio1)
    await page.update_async()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play", on_click=lambda _: audio1.play()),
    )

ft.app(target=main)

**Comment by jkloip** [comment, marked as answer]
I have solved this problem. The solution process is briefly explained as follows:
First I updated flet to **0.21.2**, then I made some changes to the program code. I removed **await page.update_async()** in **audio1**, and then used flet build apk --include-packages flet_audio_recorder flet_audio to package the program. . After this process, you can record and play the recording files on your Android phone
The entire program code is as follows：
import flet as ft
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder"), center_title=True)

    path = "test-audio-file.wav"
    url_1="MetroSVT.mp3"

    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)
    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src="MetroSVT.mp3",
        autoplay=False,
        volume=6,
        balance=0,
    )
    page.overlay.append(audio1)

    def handle_start_play(e):
        audio1.src="test-audio-file.wav"
        audio1.update()
        audio1.play()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play1-def-ply", on_click=lambda _: audio1.play()),
        ft.ElevatedButton("Play2-rec-ply", on_click=handle_start_play),
        ft.ElevatedButton("Play3-def-ply", on_click=lambda _: audio1.play()),
        ft.ElevatedButton("Play4-rec-ply", on_click=handle_start_play),
    )

ft.app(target=main)

I hope the above information can be helpful to people who encounter the same problem.
